src/app/util/ember-compatibility-functions.cpp

```diff
@@ -542,62 +586,88 @@ CHIP_ERROR prepareWriteData(const EmberAfAttributeMetadata * metadata, TLV::TLVR
+            ConcreteAttributePath path(aClusterInfo.mEndpointId, aClusterInfo.mClusterId, aClusterInfo.mAttributeId);
+            AttributeValueDecoder valueDecoder(aReader);
+            ReturnErrorOnFailure(attrOverride->Write(path, valueDecoder));
 
-    return ToInteractionModelStatus(emberAfWriteAttributeExternal(aClusterInfo.mEndpointId, aClusterInfo.mClusterId,
-                                                                  aClusterInfo.mAttributeId, CLUSTER_MASK_SERVER, 0, attributeData,
-                                                                  attributeMetadata->attributeType));
-}
+            auto accessControlStatus = CheckAccessControl(aSubjectDescriptor, aClusterInfo.mEndpointId, aClusterInfo.mClusterId, /* write = */ true);
+            if (accessControlStatus != Protocols::InteractionModel::Status::Success)
+            {
+                status = accessControlStatus;
+                ExitNow();
+            }
+            accessControlChecked = true;
 
-CHIP_ERROR WriteSingleClusterData(ClusterInfo & aClusterInfo, TLV::TLVReader & aReader, WriteHandler * apWriteHandler)
-{
-    AttributePathParams attributePathParams;
-    attributePathParams.mEndpointId  = aClusterInfo.mEndpointId;
-    attributePathParams.mClusterId   = aClusterInfo.mClusterId;
-    attributePathParams.mAttributeId = aClusterInfo.mAttributeId;
+            if (valueDecoder.TriedDecode())
+            {
+                status = Protocols::InteractionModel::Status::Success;
+                ExitNow();
+            }
+        }
 
-    // TODO: Refactor WriteSingleClusterData and all dependent functions to take ConcreteAttributePath instead of ClusterInfo
-    // as the input argument.
-    AttributeAccessInterface * attrOverride = findAttributeAccessOverride(aClusterInfo.mEndpointId, aClusterInfo.mClusterId);
-    if (attrOverride != nullptr)
-    {
-        ConcreteAttributePath path(aClusterInfo.mEndpointId, aClusterInfo.mClusterId, aClusterInfo.mAttributeId);
-        AttributeValueDecoder valueDecoder(aReader);
-        ReturnErrorOnFailure(attrOverride->Write(path, valueDecoder));
+        // Passing nullptr as buf to emberAfReadAttribute means we only need attribute type here,
+        // and ember will not do data read & copy in this case.
+        const EmberAfAttributeMetadata * metadata = emberAfLocateAttributeMetadata(
+            aClusterInfo.mEndpointId, aClusterInfo.mClusterId, aClusterInfo.mAttributeId, CLUSTER_MASK_SERVER, 0);
+        if (metadata == nullptr)
+        {
+            status = Protocols::InteractionModel::Status::UnsupportedAttribute;
+            ExitNow();
+        }
 
-        if (valueDecoder.TriedDecode())
+        if (!accessControlChecked)
         {
-            return apWriteHandler->AddStatus(attributePathParams, Protocols::InteractionModel::Status::Success);
+            auto accessControlStatus = CheckAccessControl(aSubjectDescriptor, aClusterInfo.mEndpointId, aClusterInfo.mClusterId, /* write = */ true);
```

Comment:
As before, shouldn't this be a function of the privilege specified on the attribute?

Reply:
If you look in that function, you'll see a TODO noting this. It won't happen in this PR.